### ISSUE TITLE
`redirect` from "next/navigation" does not work inside `setTimeout` in a `useEffect` body

### ISSUE THREAD

**Issue by rileybarabash** [vercel/next.js#59489]
### Link to the code that reproduces this issue

https://github.com/rileybarabash/next-bug-repro-useeffect-settimeout-redirect

### To Reproduce

1. Start in development (pnpm run dev)
2. Count to 5
3. Realize that no matter how long you wait, it is not going to redirect
4. Uncomment and try the 2 other examples in "~/page.tsx", one which omits `setTimeout` and the other that uses `useRouter.push`

### Current vs. Expected behavior

Does not redirect, expected a redirect. Very easily reproducible.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: x64
  Version: Darwin Kernel Version 23.1.0: Tue Oct  3 22:41:14 PDT 2023; root:xnu-10002.41.8~15/RELEASE_X86_64
Binaries:
  Node: 20.10.0
  npm: 10.2.3
  Yarn: N/A
  pnpm: 8.10.2
Relevant Packages:
  next: 14.0.5-canary.5
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Routing (next/router, next/navigation, next/link)

### Additional context

I found this related issue: #59163. Thoroughly read the relevant [docs](https://nextjs.org/docs/app/api-reference/functions/redirect) and I believe this should work.

**Comment by JoseVSeb** [comment]
I don't think this is a bug. `redirect` works by throwing the redirection "error" which is caught outside of the function/component by NextJS. Here, `setTimeout` takes the call to `redirect` outside of the stack such that the redirection is no longer caught by the NextJS renderer. That is not the case for `router.push`, which does not throw anything. It also the reason why we need to wrap calls to server actions in `startTransition` if we need to manually call the server action without the form `action` prop, becuase async functions are outside the stack by default and need to be awaited where NextJS can catch it, hence `startTransition`.

P.S. I'm only speaking from my understanding of how NextJS works, I could be mistaken.

**Comment by rileybarabash** [comment]
You're right, and I'm aware that it relies on throwing an unhandled error. That is probably the cause here. But to the average developer and according to the docs, this is unexpected behaviour.

I don't know much about the inner workings of `setTimeout` myself, but the fact that `redirect` silently failed in what seems like a bare-bones implementation threw me off.

My thoughts are that any errors thrown inside `setTimeout` should propagate back up to the original call, or at the bare minimum (if not rethrown inside `setTimeout`) it should print logs or error somewhere, similar to how `redirect` does in a try/catch block.

**Comment by anptrs** [comment]
This bit me too, according to docs it should work in client components, and I can confirm redirect() does not work as expected in a timer.

**Comment by JoseVSeb** [comment]
wrap the `redirect` inside `startTransition`. it should work then.

```
  const [, startTransition] = useTransition();

  useEffect(() => {
    const timeout = setTimeout(() => startTransition(() => redirect("/will-never-be-hit")), 5000);

    return () => clearTimeout(timeout);
  }, []);
```

not sure if there's any other solution, but this will make sure the redirection error is caught by NextJS. it's what is from the docs for server actions.

might as well use `router.push`. it's what should be used in client components anyway.

**Comment by rileybarabash** [comment]
moved repro [here](https://github.com/rileybarabash/bug-repros-and-demos/tree/next-useeffect-settimeout-redirect).

Thank you @JoseVSeb for the workaround. I do agree, `router.push` works great and is fine, but this is about `redirect` and contributing feedback to fuel and aid in the development of clean APIs.

Again, according to the docs, it seems like it is the perfect all-around function for programatically navigating to new routes, both server and client side, and is a much more declarative syntax for the use case.

This was clearly an overstep in the docs or the implementation, I am keeping it open for that reason.
